### PR TITLE
[uorb-graph][squash][refactor]  Fixes PR #16695 -- Fixes uorb-graph script

### PR DESCRIPTION
**Pull request by teyrana**
## Problem Description

(Updated: Mon 22 Mar 2021 09:14:00 PM EDT )

The script at `Tools/uorb_graph/create.py` was undercounting publications and subscriptions -- particularly C++ definitions.
This PR overhauls & refactors the script to greatly increase the variety of cases it can handle.

## Describe your solution
This adds additional fields to the regexes when scanning source code.
- Expands the debugging options of this script
    - adds '-v', '-vv' => increase output verbosity
    - adds '--merge-depends' flag for merging library topics
- Records all warnings -- unhandled or ambiguous matches
    - A simple summary is printed on completion
    - A listing is printed out in verbose/debug output modes
    - only prints out for files found in the supplied source tree (i.e. the `--src-path` argument)
- Library vs Module
    - By default, modules don't link to their dependent libraries.
    - Dependency libraries may be merged with their modules with the `--merge-depends` CLI flag.
- Overhauls the regexes used to detect matches
    - Expands several of the regexes to match more topic declarations

## Test data / coverage

This patch only affects a developer documentation script, and has no effect on flight code

During development, all test cases were run from the path: `/home/teyrana/project/px4/Tools/uorb_graph`

### Before

Run on `master` branch @ 9d47f7ecda

```
$ ./create.py -o graphviz -s ../../src/
Excluded topics: ['parameter_update', 'mavlink_log', 'log_message']
...
number of modules: 33
number of topics: 5
...
```

### Examples + After Results

1. Module: `rc_input`
    - Status: Fixed
    - `src/drivers/rc_input/RCInput.cpp`:264:`uORB::Publication<vehicle_command_ack_s> vehicle_command_ack_pub{ORB_ID(vehicle_command_ack)};`
    - `src/drivers/rc_input/RCInput.hpp`:146:`uORB::Subscription	_vehicle_cmd_sub{ORB_ID(vehicle_command)};`
    - `src/drivers/rc_input/RCInput.hpp`:155:`uORB::PublicationMulti<input_rc_s>	_to_input_rc{ORB_ID(input_rc)};`
    - warning sites: 0
    - ambiguous topics: 0
    - topic count: 8

2. Module: `iridiumsbd`
    - Status: Fixed
    - `src/drivers/telemetry/iridiumsbd/IridiumSBD.h`:304:
        `uORB::Publication<iridiumsbd_status_s> _iridiumsbd_status_pub{ORB_ID(iridiumsbd_status)};`
    - warning count: 0
    - topic count: 1

3.  Module `frsky_telemetry`
    - Status: Fixed
    - example: `src/drivers/telemetry/frsky_telemetry/frsky_data.cpp`:69:`uORB::SubscriptionData<vehicle_acceleration_s> vehicle_acceleration_sub{ORB_ID(veh...`
    - warning count: 1
    - topic count: 7

4. Module `navigator`
    - Status: Fixed
    - `src/modules/navigator/geofence.h`:187 `uORB::SubscriptionData<vehicle_air_data_s>	_sub_airdata;`
    - `src/modules/navigator/mission_block.h`:155:`uORB::Publication<actuator_controls_s>	_actuator_pub{ORB_ID(actuator_controls_2)};`
    - `src/modules/navigator/navigator_main.cpp`:103:`_local_pos_sub = orb_subscribe(ORB_ID(vehicle_local_position));`
    - `src/modules/navigator/gpsfailure.h`:79:`uORB::Publication<vehicle_attitude_setpoint_s>	_fw_virtual_att_sp_pub{ORB_ID(fw_virtual_attitude_setpoint)};`
    - `src/modules/navigator/mission.h`:L246:`	uORB::Publication<navigator_mission_item_s> _navigator_mission_item_pub{ORB_ID::navigator_mission_item};`
    - Ambiguous lines: 1
    - Ambiguous topics: 1
    - Topics: 24

5. Module: `temperature_compensation`
    - Status: Fixed
              Overcounts lines as warnings which aren't strictly ambiguous, but are harder to process
    - `src/modules/temperature_compensation/TemperatureCompensationModule.h`:122:`uORB::Subscription _baro_subs[BARO_COUNT_MAX]`
    - warning sites: 18
    - warning topics: 0
    - topic count: 7

6. Module: `commander`
    - Status: Fixed
    - `src/modules/commander/Commander.hpp`:428: `uORB::SubscriptionMultiArray<battery_status_s, battery_status_s::MAX_INSTANCES> _battery_status_subs{ORB_ID::battery_status};`
    - `src/modules/commander/level_calibration.cpp`:85: `uORB::SubscriptionBlocking<vehicle_attitude_s> att_sub{ORB_ID(vehicle_attitude)};`
    - warning count:  20
    - warnings topics: 0
    - topic count: 45

7. Module: `vtol_att_control`
    - Status: Fixed
    - `src/modules/vtol_att_control/vtol_att_control_main.h`:135:`uORB::SubscriptionCallbackWorkItem _actuator_inputs_fw{this, ORB_ID(actuator_controls_virtual_fw)};`
    - `src/modules/vtol_att_control/vtol_att_control_main.h`:154:`uORB::Publication<actuator_controls_s>		_actuators_0_pub{ORB_ID(actuator_controls_0)};`
    - warning count:  0
    - warnings topics: 0
    - topic count: 20

8. Module: `airspeed_selector`
    - Status: Fixed
    - `src/modules/airspeed_selector/airspeed_selector_main.cpp`:L122:`uORB::SubscriptionMultiArray<airspeed_s, MAX_NUM_AIRSPEED_SENSORS> _airspeed_subs{ORB_ID::airspeed};`
    - Ambiguous lines: 0
    - Ambiguous topics: 0
    - Topics: 13

9. Module: `mavlink`
    - Status: Fixed
    - `src/modules/mavlink/streams/ACTUATOR_CONTROL_TARGET.hpp`:79:`_act_ctrl_sub = new uORB::Subscription{ORB_ID(actuator_controls_0)};`
    - Ambiguous lines: 5
    - Ambiguous topics: 1
    - Topics: 25

10. Module (Library): `drivers__hott`
    - Status: Fixed
    - `src/drivers/telemetry/hott/messages.cpp`:78:`_home_sub = orb_subscribe(ORB_ID(home_position));`
    - Warning sites: 6
    - Warning topics: 6
    - Topics: 6

11. Module `hott_telemetry`
    - Status: Fixed
    - references 'drivers__hott' library in local parent directory...and  _that_ library contains messages
    - without '-c','--combine-dependencies'
        - warning count: 0
        - topic count: 0   (all topics are defined in 'dependency' library)
    - with '-c','--combine-dependencies'
        - warning count: 6
        - topic count: 6

12. Module Nested in another Modules
    - for the nested modules 'simulator' => 'battery_simulator': topics go from 17 => 0
    - how common are nested modules? ... Here's the list:
        - drivers__hott > hott_telemetry    # nested AND dependency-linked
        - drivers__hott > hott_sensors      # nested but unlinked
        - lightware_laser_serial > lightware_laser_test
        - bmi088 > bmi088_i2c
        - tests > hrt_test
        - controllib > controllib_test
        - cdev > cdev_test
        - mavlink > mavlink_tests
        - uorb_msgs_microcdr ${uorb_sources_microcdr}) > micrortps_client
        - commander > commander_tests
        - simulator > battery_simulator

### After: entire source tree

1. Modules: all
    - Scope count:     90
    - Topics:         105
    - Warning lines:  158

2. Modules + Libraries: (w/union):
    - Scope count:     96
    - Topics:         168
    - Warning lines:  158

3.  Modules + Libraries (w/merge):
    - Scope count: 162
    - Topic count: 170
    - Warning count: 158

### CODE REVIEW

**Comment by teyrana** [comment]
Edit: unborked the rebase.

**Comment by teyrana** [comment]
@LorenzMeier @bkueng 
Looks ready for review and/or merge.

File: Tools/uorb_graph/create.py [hunk 1]
```diff
@@ -102,49 +103,39 @@ def match(self, source_line: str) -> Set[str]:
...
                     log.debug("            >>> Found ORB_ID topic: " + topic_group +  "    w/regex: " + str(regex.pattern))
-                    self._add_topic(matches, topic_group)
-                    break
-                elif route_group == '<' and topic_group.endswith('_s'):
-                    topic_group = topic_group[:-2]
-                    log.debug("            >>> Found C++ template-declaration: " + topic_group +  "    w/regex: " + str(regex.pattern))
-                    self._add_topic(matches, topic_group)
-                    # continue processing
+                    return self._filter_topic(topic_group)
                 elif route_group in ['{','('] and topic_group.endswith('_s'):
```
Comment (bkueng) [inline Tools/uorb_graph/create.py:107]: This can be removed as well (?)

File: Tools/uorb_graph/create.py [hunk 2]
```diff
@@ -102,49 +103,39 @@ def match(self, source_line: str) -> Set[str]:
...
-                        log.debug("            >>> Found array declaration: " + topic_group +  "    w/regex: " + str(regex.pattern))
-                        self._add_topic(matches, topic_group)
-                        break
-                    else:
-                        # no topic found -- ambiguity -- return an empty set
-                        return set()
+                    if not topic_group:
+                        log.debug("            !! found an ambiguous site => return an empty set")
+                        return PubSub.AMBIGUOUS_SITE_TOPIC
                 elif 'Multi' in route_group and topic_group.endswith('_s'):
```
Comment (bkueng) [inline Tools/uorb_graph/create.py:115]: And this too

**Review by bkueng — COMMENTED** [review]
Thanks, this is coming together nicely.

Observerations:
- if not merge depends and no scope whitelist, the result does not contain the libraries
- `-['mavlink']:src/modules/mavlink/streams/ACTUATOR_CONTROL_TARGET.hpp         :79 =        _act_ctrl_sub = new uORB::Subscription{ORB_ID(actuator_controls_0)};` I think this warning case we can easily handle as well. The rest looks as I would expect.

Next we can start to deploy this again to the docs.

File: Tools/uorb_graph/create.py (hunk 1 above)
Comment (teyrana) [reply Tools/uorb_graph/create.py:107]: oops. yea.

Comment (teyrana) [reply Tools/uorb_graph/create.py:107]: (removed in local branch.)

File: Tools/uorb_graph/create.py (hunk 2 above)
Comment (teyrana) [reply Tools/uorb_graph/create.py:115]: Ditto. (removed in local branch.)

**Comment by teyrana** [comment]
look into that false negative line...

**Comment by teyrana** [comment]
The CI-only issue is back.  Troubleshooting.

**Comment by teyrana** [comment]
... well the code is working correctly.  It's finding topics in `src/lib/parameters/` which is outside any detected scope. 

Are these "legit" call sites?

**Comment by bkueng** [comment]
> ... well the code is working correctly. It's finding topics in src/lib/parameters/ which is outside any detected scope.

Interesting, this is a unit test code which we can ignore for the graph. `px4_add_functional_gtest` defines that scope (ignoring that file specifically is fine though).

**Comment by teyrana** [comment]
Oh, that's not the only place it exists... Just the first one.

I'll see how hard it is to detect a test scope...

**Comment by teyrana** [comment]
Okay, given the uorb-graph check is no longer in the list, I guess this is ready to merge.

**Review by bkueng — APPROVED** [review]

**Pull request merged**